# **Multiple Linear Regression**

**Import the relevant libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression

**Load the data**

In [2]:
data = pd.read_csv("dataset/Multiple_linear_regression.csv")
data

SAT   GPA  Rand 1,2,3
0   1714  2.40           1
1   1664  2.52           3
2   1760  2.54           3
3   1685  2.74           3
4   1693  2.83           2
..   ...   ...         ...
79  1936  3.71           3
80  1810  3.71           1
81  1987  3.73           3
82  1962  3.76           1
83  2050  3.81           2

[84 rows x 3 columns]

In [3]:
data.describe()

SAT        GPA  Rand 1,2,3
count    84.000000  84.000000   84.000000
mean   1845.273810   3.330238    2.059524
std     104.530661   0.271617    0.855192
min    1634.000000   2.400000    1.000000
25%    1772.000000   3.190000    1.000000
50%    1846.000000   3.380000    2.000000
75%    1934.000000   3.502500    3.000000
max    2050.000000   3.810000    3.000000

This dataset contains 84 samples which we get from count

Samples is the machine learning term for observations

**Create the multiple linear regression**

**Declare the dependent and independent variables**

In [4]:
x = data[['SAT','Rand 1,2,3']]
y = data['GPA']

**Regression itself**

In [5]:
reg = LinearRegression()
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [6]:
reg.coef_

array([ 0.00165354, -0.00826982])

The output is arranged in the way we feed them

SAT = 0.00165354

Rand 1,2,3 = -0.00826982

In [7]:
reg.intercept_

0.29603261264909486

**Calculating the R-squared**

The R-squared is a universal measure to evaluate how well linear regressions fare and compare

In [8]:
reg.score(x,y) # r-squared value

0.4066811952814285

The Adjusted R-squared steps on the R-squared and adjusts for the number of variables included in the model

Formula for Adjusted R^2

$R^2_(adj.) = 1 - (1-R^2)*\frac(n-1){n-p-1}$


In [9]:
x.shape

(84, 2)

n = 84(the number of observations)


p =2(the number of predictors)

As sklearn doesn't have a method to display Adjusted R squared you have to create the formula in Python

In [10]:
r2 = reg.score(x,y)

n = x.shape[0]

p = x.shape[1]

adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134023

**Feature Selection** simplifies models, improves speed and prevents a series of unwanted issues arising from having too many features

feature_selection.f_regression

F-regression creates simple linear regressions of each feature and the dependent variables

1. GPA <- SAT
2. GPA <- Rand 1,2,3

Note: for simple linear regression, the p-value of F-stat = the p-value of the only independent variable

In [11]:
from sklearn.feature_selection import f_regression

In [12]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

The first array is F-statistics
and second is p-values

In [13]:
p_values = f_regression(x,y)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

e-11 = *10 ^-11 = /10^11
e-1 = *10^-1 = /10


In [14]:
p_values.round(3)

array([0.   , 0.676])

Note: these are the univariate p-values reached from simple linear models. They do not reflect the interconnection of the features in our mutliple linear regression

**Creating a summart table**

In [15]:
reg_summary = pd.DataFrame(data =['SAT','Rand 1,2,3'], columns=['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [16]:
reg_summary['Coefficients']=reg.coef_
reg_summary['p-values'] = p_values.round(3)
reg_summary

Features  Coefficients  p-values
0         SAT      0.001654     0.000
1  Rand 1,2,3     -0.008270     0.676

p-values are one of the best ways to determine if a variable is redundant, but they provide no information whatsoever about HOW USEFul a variable is

**Standardization**

In [17]:
from sklearn.preprocessing import StandardScaler

In [18]:
scaler = StandardScaler()

In [19]:
scaler.fit(x)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [20]:
x_scaled = scaler.transform(x)

In [21]:
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

**Regression with scaled features**

In [22]:
reg = LinearRegression()
reg.fit(x_scaled,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [23]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [24]:
reg.intercept_

3.330238095238095

**Creating a summary table**

In [25]:
reg_summary = pd.DataFrame([['Bias'],['SAT'],['Rand 1,2,3']], columns = ['Features'])
reg_summary['Weights'] = reg.intercept_,reg.coef_[0],reg.coef_[1]

In [26]:
reg_summary

Features   Weights
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

Weights is the machine learning word for co-efficients

The ML word for intercept is 'bias'

The closed a weight is to 0, the smaller its impact; The bigger the weight the bigger its impact

**Making predictions with the standardized co-efficients(weights**

In [27]:
new_data = pd.DataFrame(data = [[1700,2],[1800,1]], columns = ['SAT','Rand 1,2,3'])
new_data

SAT  Rand 1,2,3
0  1700           2
1  1800           1

In [28]:
reg.predict(new_data)

array([295.39979563, 312.58821497])

In [29]:
new_data_scaled = scaler.transform(new_data)
new_data_scaled

array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

In [30]:
reg.predict(new_data_scaled)

array([3.09051403, 3.26413803])

**What if we removed the 'Random 1,2,3' variable?**

In [31]:
reg_simple = LinearRegression()
x_simple_matrix = x_scaled[:,0].reshape(-1,1)
reg_simple.fit(x_simple_matrix,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [32]:
reg_simple.predict(new_data_scaled[:,0].reshape(-1,1))

array([3.08970998, 3.25527879])